### PR TITLE
Refactor discover function

### PR DESCRIPTION
**Pull request by sz-piotr**
This PR continues the work of refactoring discovery to use dependency injection and add tests.

This time it concerns itself with the `discover` function and the core discovery loop.

### CODE REVIEW

File: packages/backend/src/core/discovery/engine/DiscoveryEngine.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  async discover(config: DiscoveryConfig) {
+    const resolved: Analysis[] = []
+    const stack = new DiscoveryStack()
+    stack.push(config.initialAddresses, 0)
+
+    while (!stack.isEmpty()) {
+      const item = stack.pop()
+
+      const reason = shouldSkip(item, config)
```
Comment (torztomasz) [inline packages/backend/src/core/discovery/engine/DiscoveryEngine.ts:21]: can't we put it as a method in DiscoveryStack?

**Review by torztomasz — COMMENTED** [review]
The only thing that looks kinda ugly is keeping discover.ts and runDiscover.ts, can't we put theirs logic inside DiscoveryRunner?

File: packages/backend/src/core/discovery/engine/DiscoveryEngine.ts (hunk 1 above)
Comment (sz-piotr) [reply packages/backend/src/core/discovery/engine/DiscoveryEngine.ts:21]: It's a pure function, I don't see a particular need.

**Comment by sz-piotr** [comment]
> The only thing that looks kinda ugly is keeping discover.ts and runDiscover.ts, can't we put theirs logic inside DiscoveryRunner?

Yeah, the plan was to refactor bottom up, so I started a few PRs ago with AddressAnalyzer and so on, got to DiscoveryEngine and I'll keep going up. It's just a lot to do in one step, so its multiple PRs

**Review by torztomasz — APPROVED** [review]

**Pull request merged**